File: inference/MeanVarNullFilter.m
%CLASS: MEAN VARIANCE NULL FILTER
%See superclass EmpiricalNullFilter
%Does the empirical null filter, makes use of ImageJ and multiple threads
%Replaces empirical null mean with mean
%Replaces empirical null var with variance
classdef MeanVarNullFilter < EmpiricalNullFilter
  
  methods (Access = public)
    
    %CONSTRUCTOR
    %PARAMETERS:
      %radius: the radius of the kernel
    function this = MeanVarNullFilter(radius)
      this@EmpiricalNullFilter(radius);
      this.javaFilter = uk.ac.warwick.sip.empiricalnullfilter.MeanVarNullFilter();
      this.javaFilter.setRadius(radius);
    end
    
  end
  
end

File: java/src/uk/ac/warwick/sip/empiricalnullfilter/MeanVarNullFilter.java
//MIT License
//Copyright (c) 2019 Sherman Lo

package uk.ac.warwick.sip.empiricalnullfilter;

import org.apache.commons.math3.distribution.NormalDistribution;
import org.apache.commons.math3.random.RandomGenerator;

//CLASS: MEAN VAR NULL FILTER
/**Subclass of EmpiricalNullFilter, replaces the empirical null mean with mean, replaces the 
 *     empirical null variance with variance.
 * @author Sherman Lo
 */
public class MeanVarNullFilter extends EmpiricalNullFilter {
  
  public MeanVarNullFilter() {
    super();
    //this filter does not need to copy kernel pixels or work out quantiles
    this.isKernelCopy = false;
    this.isKernelQuartile = false;
  }
  
  @Override
  protected float[] getNullMeanStd(float initialValue, Kernel kernel, NormalDistribution normal,
      RandomGenerator rng) {
  //declare 2 vector to store the null mean and null std
    float[] nullMeanStd = new float[2];
    //get the empirical null
    nullMeanStd[0] = kernel.getMean();
    nullMeanStd[1] = kernel.getStd();
    return nullMeanStd;
  }
}
